### ISSUE TITLE
tdData needs to pick up more edge cases

### ISSUE THREAD

**Issue by NattyE** [NattyE/nlstimedist#7]
Is it possible that the data could contain NAs, for example? @NicSteer do let me know if you come up with any more bugs with this function.

**Comment by NattyE** [comment]
@NicSteer to avoid complications with authorization tokens I have now made the repo public. You can see the latest (easier) installation instructions here: https://github.com/NattyE/nlstimedist/blob/develop/README.md

To install the develop branch, add the argument `ref = "develop"`

**Comment by NicSteer** [comment]
I have installed the package as per instructions and have begun testing the tdData function.  Using the example Aphid code from above it is still producing this error:  

> Error in tdData(aphid, x = "x", y = "y", runNo = "year") : 
>    unused argument (runNo = "year")

After removing the runNo argument it is able to create the tmp object.  Perhaps more worrying though is that the function is producing incorrect parameter values:
> y1994 <- tmp$clean %>% filter(year == 1994)
> timedist(y1994, x = "x", y = "y", r = 0.01, c = 0.05, t = 200)
Nonlinear regression model
  model: y ~ 1 - (1 - (r/(1 + exp(-c * (x - t)))))^x
   data: data
        r         c         t 
  0.07957   0.15502 132.25087 
 residual sum-of-squares: 4981075

> Number of iterations to convergence: 8 
Achieved convergence tolerance: 1.49e-08

I have run this data previously with the timedist function and have tested it against the SPSS output also and both of these produce the below parameter values:
> lmaphidN <- timedist(aphidN, x = "x", y = "y1994", r = 0.01, c = 0.2, t = 200)
> lmaphidN
Nonlinear regression model
  model: y1994 ~ 1 - (1 - (r/(1 + exp(-c * (x - t)))))^x
   data: data
        r         c         t 
8.025e-03 2.656e-01 1.950e+02 
 residual sum-of-squares: 0.2293

> Number of iterations to convergence: 7 
Achieved convergence tolerance: 1.49e-08

SPSS: r = 0.008, c = 0.266, t = 194.965

The timedist model using the tdData cleaned data will not start with the starting values used above and with SPSS:
> timedist(y1994, x = "x", y = "y", r = 0.01, c = 0.2, t = 200)
Error in assertr::verify(vars, c > 0) : verification failed! (1 failure)

I have tested tdData cleaned data against the same data that has been manually set up for the timedist function (proportions worked out) from different datasets. With the pre-worked out data, timedist is producing parameter values that mirror SPSS but when using the cleaned data it is producing errors which are usually associated with incorrect starting values (even when the starting values are known to be correct):
>  Error in nlsModel(formula, mf, start, wts) : 
      singular gradient matrix at initial parameter estimates

After further investigation.....

The function seems to be picking out the right number of values and working out cumN but it is not calculating the proportions correctly from this maximun value (in the example below 16).  This helps to explain why it is not calculating the correct parameter values.

> print(y9.8)
  Day temp y cumN     propMax
1  18  9.8 3    3 0.009146341
2  19  9.8 3    6 0.018292683
3  20  9.8 1    7 0.021341463
4  21  9.8 1    8 0.024390244
5  22  9.8 1    9 0.027439024
6  23  9.8 2   11 0.033536585
7  25  9.8 1   12 0.036585366
8  28  9.8 2   14 0.042682927
9  29  9.8 2   16 0.048780488

This is the same reason for the disparity between the y1994 parameter values produced for the Aphid dataset.  The function has correctly worked out the cumN but has produced incorrect proportions.

> print(y1994)
      x  X year    y cumN      propMax
1   140 NA 1994    1    1 1.031406e-05
2   144 NA 1994    1    2 2.062813e-05
3   145 NA 1994    1    3 3.094219e-05
4   152 NA 1994    3    6 6.188438e-05
5   153 NA 1994    3    9 9.282657e-05
.
.
.
120 311 NA 1994    2 8953 9.234181e-02
121 312 NA 1994    2 8955 9.236244e-02
122 313 NA 1994    2 8957 9.238306e-02
123 314 NA 1994    1 8958 9.239338e-02
124 315 NA 1994    1 8959 9.240369e-02

Going forward.....

We have two options: (i) amend the tdData function so that it correctly calculates the proportions from the maximum value of cumN and I will continue testing or (ii) due to the time you have already put into this, we abandon the tdData parser function all together (which may seem drastic) and ask users to have the data set up correctly ready to use the timedist function.

Please let me know your thoughts and we will proceed accordingly.

**Comment by NattyE** [comment]
@NicSteer I have looked into the issues you are seeing separately. 
TL;DR: I don't think it would be the best idea to scrap `tdData`.
Here is some help for each:

# Issue 1:
I think you have some errors in your code. Please try running the following and let me know how you get on

    # Load in the aphid data
    aphid <- read.csv("tdAphid.csv")
    library(tidyr)
    library(dplyr)
    # The data are not in tidy format. We correct this below:
    aphidTidy <- 
      aphid %>% 
      gather(key = year, value = y, y1994:y2014) %>% 
      mutate(year = as.numeric(gsub("y", "", x = .$year)))
    
    # Install the development version of the package and load it
    devtools::install_github("NattyE/nlstimedist", ref = "develop")
    library(nlstimedist)
    
    # Clean the data
    cleanAphid <- tdData(aphidTidy, x = "x", y = "y", group = "year")
    
    # Filter out the 1994 data
    y1994 <- cleanAphid$clean %>% 
      filter(year == 1994)
    
    # Run the model
    # Note: runYear has been changed to group to be more generic
    timedist(data = y1994, x = "x", y = "cumN", r = 0.01, c = 0.2, t = 200)

# Issue 2:
When you see this error:

    timedist(y1994, x = "x", y = "y", r = 0.01, c = 0.2, t = 200)
    Error in assertr::verify(vars, c > 0) : verification failed! (1 failure)

It is coming from the moment generating functions - not from the `timedist` function. Essentially the algorithm has produced a value of `c` which is `< 0` and so this has been flagged as failing since the moment generating functions use the same value checking tests for `r`, `c` and `t` as the `timedist` function does. Is it possible for the model to produce valid values beyond the following?

    0 < r <= 1, c >0, t >= 0

If not, then we need to take a look at the moment generating functions. i.e. The model is being produced with the starting values you are using, but it fails when trying to calculate the moments. Admittedly this is poor since the warning message produced is exactly the same and doesn't pinpoint which function it is coming from.

# Issue 3:
Going forwards I think we should continue with the `tdData` function. I think it does produce the correct values. We now need to look more closely at the moment generating functions.

**Comment by NicSteer** [comment]
My first attempt at this I copy and pasted the code from your previous post with the only difference being that the runNo kept producing an error so maybe that's why it didn't work as it should.

This time I have copy and pasted the chuck of code from Issue 1 above and the function is still producing incorrect parameter values. The proportions have been worked out correctly by tdData and match exactly what I have but the parameter values calculated by timedist differ from those produced by timedist from the pre-calculated proportions and SPSS.

tdData output:
> aphid <- read.csv("Aphid.csv", header = TRUE, sep = ",")
> aphidTidy <- 
   aphid %>% 
   gather(key = year, value = y, y1994:y2014) %>% 
   mutate(year = as.numeric(gsub("y", "", x = .$year)))
> library(nlstimedist)
> cleanAphid <- tdData(aphidTidy, x = "x", y = "y", group = "year")
Warning message:
In tdData(aphidTidy, x = "x", y = "y", group = "year") :
  Replaced 16 NAs with 0
> y1994 <- cleanAphid$clean %>% 
   filter(year == 1994)
> timedist(data = y1994, x = "x", y = "cumN", r = 0.01, c = 0.2, t = 200)
Nonlinear regression model
  model: cumN ~ 1 - (1 - (r/(1 + exp(-c * (x - t)))))^x
   data: data
        r         c         t 
  0.08964   0.04432 161.20638 
 residual sum-of-squares: 5.581e+09

> Number of iterations to convergence: 2 
Achieved convergence tolerance: 1.49e-08
Warning messages:
1: In if (class(data) == "td") data <- data$clean :
  the condition has length > 1 and only the first element will be used
2: In mean.default(as.list(model$m$getEnv())$y) :
  argument is not numeric or logical: returning NA

Pre-calculated proportion output (exact same proportions as tdData calculates):
> aphidN <- read.csv("AphidNewc.csv", header = TRUE, sep = ",")
> lmaphidN <- timedist(aphidN, x = "x", y = "y1994", r = 0.01, c = 0.2, t = 200)
> lmaphidN
Nonlinear regression model
  model: y1994 ~ 1 - (1 - (r/(1 + exp(-c * (x - t)))))^x
   data: data
        r         c         t 
8.025e-03 2.656e-01 1.950e+02 
 residual sum-of-squares: 0.2293

> Number of iterations to convergence: 7 
Achieved convergence tolerance: 1.49e-08

SPSS output:
r = 0.008, c = 0.266, t = 194.965

As you can see the pre-calculated and SPSS outputs match but the timedist output using the tdData function produces totally different parameter values.  I have tested this with additional Aphid data:

> y1996 <- cleanAphid$clean %>% 
   filter(year == 1996)
> timedist(data = y1996, x = "x", y = "cumN", r = 0.05, c = 0.05, t = 300)
Error in nlsModel(formula, mf, start, wts) : 
  singular gradient matrix at initial parameter estimates
In addition: Warning message:
In if (class(data) == "td") data <- data$clean :
  the condition has length > 1 and only the first element will be used

Pre-calculated proportion output (exact same proportions as tdData calculates)
> lmaphidN96 <- timedist(aphidN, x = "x", y = "y1996", r = 0.05, c = 0.05, t = 300)
> lmaphidN96
Nonlinear regression model
  model: y1996 ~ 1 - (1 - (r/(1 + exp(-c * (x - t)))))^x
   data: data
        r         c         t 
9.351e-03 6.768e-02 2.575e+02 
 residual sum-of-squares: 0.8105

> Number of iterations to convergence: 16 
Achieved convergence tolerance: 1.49e-08

SPSS ouput:
r = 0.009, c = 0.068, t = 257.488

Again the pre-calculated method and SPSS match but this time the tdData sorted method is unable to start even with the same starting parameters that work for the other two methods.  This is the same for the data from 1999 also.

I have identified similar issues with the Lobelia dataset as well.  I will continue testing tomorrow and try and give you more information.  If you notice that I have done anything incorrectly please let me know.

**Comment by NattyE** [comment]
So I think that you maybe gave the function the wrong column (sorry this is my fault as I gave you incorrect code in the previous comment). 

    timedist(data = y1994, x = "x", y = "propMax", r = 0.01, c = 0.2, t = 200)
        Nonlinear regression model
      model: propMax ~ 1 - (1 - (r/(1 + exp(-c * (x - t)))))^x
       data: data
            r         c         t 
    8.025e-03 2.656e-01 1.950e+02 
     residual sum-of-squares: 0.2293
    
    Number of iterations to convergence: 7 
    Achieved convergence tolerance: 1.49e-08

The above line of code gives the same output as SPSS right?

**Comment by NicSteer** [comment]
I have tested this against a number of different datasets and using the propMax column the model produces accurate parameter estimates.  There is the slightest variation between some of the parameters, standard errors and confidence intervals but I am pretty sure this is just due to rounding differences between the different methods.

Just a quick query.....if I was to use the alternate method of extracting a group from a dataset using:
> y1994 <- tdData(aphid[, 1:2], x = "x", y = "y1994")

for example say I wanted the 1st and 6th columns [ , 1??] instead of the 1st and 2nd [ , 1:2] what would the notation be?

**Comment by NattyE** [comment]
Ok I'm glad it is working. There will be subtle differences. I'm not aware of any rounding that I am doing, but I suspect there will be differences between algorithms and/or programs. 

If you wanted to used the grouped version, the data would need to be "tidied" using tidyr or something similar. There is an example above using `gather`. Sorry, I'm not at my computer at the moment but I'll post a solution when I can.

**Comment by NattyE** [comment]
I think I can close this now.
